cpp/src/arrow/dataset/filter.cc

```diff
@@ -972,5 +868,33 @@ Result<std::shared_ptr<DataType>> FieldExpression::Validate(const Schema& schema
   return null();
 }
 
+RecordBatchIterator FilterBatches(RecordBatchIterator unfiltered,
```

Comment:
I think it's time to refactor the filter: RecordBatch -> RecordBatch into a separate interface. 

- Remove Expression::Evaluate. The expression class shouldn't know about RecordBatch.
- Move the FilterBatches as the basic implementation of this new class. The interface shouldn't know about Iterator<RecordBatch>, just RecordBatch -> RecordBatch.

An example:
```
// Find a better name
class FilterOperator {
    virtual Result<std::shared_ptr<RecordBatch> Filter(const std::shared_ptr<RecordBatch>& in) = 0;
    std::shared_ptr<Expression> predicate() const { return predicate_; }
  protected:
    std::shared_ptr<Expression> predicate_;
}

class StaticComputeKernelFilter : FilterOperator;
class GandivaFilter: FilterOperator;
```

This way the scheduling class, e.g. Scanner, wraps the ScanTaskIterator with a new ScanTask which applies an transparent FilterOperator. This way, we can toggle in the scanner builder (gandiva or compute, filter record batch or not).

Reply:
I disagree. IMHO it makes more sense to have filtering logic (both predicate push down via `Assume` and other via `Evaluate`) for a given expression type resident in the class which corresponds to that expression type. I'd implement evaluation against Gandiva as an overload of `Evaluate`.

The alternative involves a proliferation of switch statements on `ExpressionType` outside of `filter.cc` (or equivalently adding a visitor interface for `Expression` trees). If we can avoid that by keeping `Evaluate` and `Assume` as methods then that seems the better design to me.

Reply:
On top of separation of concerns, there's also a handful of other implementation we will explore:

- CachingOperator (caching selection vector of sub-expressions), e.g. database cracking.
- AccountingOperator a bit like MemoryPool statistics
- Reordering evaluation of predicates to exploit statistics on predicate

These optimizations/utilities have no reason living under Expression implementation. Moving this in a concrete class is a first step to enable seamless gandiva integration.